File: Data Structures & Algorithms/max-area-of-island/submission-0.cpp
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        size_t ROWS = grid.size();
        size_t COLS = grid[0].size();
        int result = 0;

        function<int(int, int)> dfs = [&](int r, int c) -> int {
            if ((r >= ROWS) || (r < 0) ||
                (c >= COLS) || (c < 0) || 
                (grid[r][c] != 1)
            ) return 0;

            grid[r][c] = -1;

            return 1 + dfs(r+1, c) + dfs(r-1, c) + dfs(r, c+1) + dfs(r, c-1);
        };

        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (grid[r][c] == 1) {
                    result = max(result, dfs(r, c));
                };
            };
        };

        return result;
    }
};

File: Data Structures & Algorithms/count-number-of-islands/submission-0.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        count = 0
        ROWS = len(grid)
        COLS = len(grid[0])
        
        def dfs(r, c):
            if r >= ROWS or r < 0 or c >= COLS or c < 0 or grid[r][c] != "1":
                return
            
            grid[r][c] = "U"

            dfs(r+1, c)
            dfs(r-1, c)
            dfs(r, c+1)
            dfs(r, c-1)

        for r, row in enumerate(grid):
            for c, cell in enumerate(row):
                if cell == "1":
                    count += 1
                    dfs(r, c)

        return count